qlib/rl/order_execution/simulator_qlib.py

```diff
@@ -2,3 +2,370 @@
+        self._twap_price = twap_price
+
+        metric_keys = list(SAOEMetrics.__annotations__.keys())  # pylint: disable=no-member
+        # NOTE: can empty dataframe contain index?
+        self.history_exec = pd.DataFrame(columns=metric_keys).set_index("datetime")
+        self.history_steps = pd.DataFrame(columns=metric_keys).set_index("datetime")
+        self.metrics = None
+
+    def update(
+        self,
+        inner_executor: BaseExecutor,
+        inner_strategy: DecomposedStrategy,
+        done: bool,
+        all_indicators: dict,
+    ) -> None:
+        execute_order = inner_strategy.execute_order
+        execute_result = inner_strategy.execute_result
+        exec_vol = np.array([e[0].deal_amount for e in execute_result])
+        num_step = len(execute_result)
+
+        if num_step == 0:
+            market_volume = np.array([])
+            market_price = np.array([])
+            datetime_list = pd.DatetimeIndex([])
+        else:
+            market_volume = np.array(
+                inner_executor.trade_exchange.get_volume(
+                    execute_order.stock_id,
+                    execute_result[0][0].start_time,
+                    execute_result[-1][0].start_time,
+                    method=None,
+                ),
+            )
+
+            trade_value = all_indicators["1min"].iloc[-num_step:]["value"].values
+            deal_amount = all_indicators["1min"].iloc[-num_step:]["deal_amount"].values
+            market_price = trade_value / deal_amount
+
+            datetime_list = all_indicators["1min"].index[-num_step:]
+
+        assert market_price.shape == market_volume.shape == exec_vol.shape
+
+        self.history_exec = _dataframe_append(
+            self.history_exec,
+            self._collect_multi_order_metric(
+                order=self._order,
+                datetime=datetime_list,
+                market_vol=market_volume,
+                market_price=market_price,
+                exec_vol=exec_vol,
+                pa=all_indicators["30min"].iloc[-1]["pa"],
+            ),
+        )
+
+        self.history_steps = _dataframe_append(
+            self.history_steps,
+            [
+                self._collect_single_order_metric(
+                    execute_order,
+                    execute_order.start_time,
+                    market_volume,
+                    market_price,
+                    exec_vol.sum(),
+                    exec_vol,
+                ),
+            ],
+        )
+
+        if done:
+            self.metrics = self._collect_single_order_metric(
+                self._order,
+                self._tick_index[0],  # start time
+                self.history_exec["market_volume"],
+                self.history_exec["market_price"],
+                self.history_steps["amount"].sum(),
+                self.history_exec["deal_amount"],
+            )
+
+        # Do this at the end
+        self.position -= exec_vol.sum()
```

Comment:
Do we have any reason to maintain this extra variable `position` instead of getting it directly from Account?

Reply:
Added a TODO in the code.